### PR TITLE
Improve type checking in getContactPhone and use CRM_Utils_Request::r…

### PR DESCRIPTION
**Pull request by seamuslee001**
…etrieve to get data from GET

Overview
----------------------------------------
This improves the type checking on the `name` key by making it use the type `Alphanumeric` rather than string. It also switches to use `CRM_Utils_Request::retrieve` rather than `CRM_Utils_Array::value` to get data from the $_GET variable

ping @eileenmcnaughton @monishdeb @colemanw

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
I tested with 

http://dmaster.local/civicrm/ajax/checkphone?cid=96,97,95,148

http://dmaster.local/civicrm/ajax/checkphone?name=Jacobs, Elbert

& this didn't work. For the form url the following patch worked

```
:...skipping...
diff --git a/CRM/Contact/Page/AJAX.php b/CRM/Contact/Page/AJAX.php
index 8a4ffdf6e8..24ffa88ae2 100644
--- a/CRM/Contact/Page/AJAX.php
+++ b/CRM/Contact/Page/AJAX.php
@@ -472,7 +472,7 @@ LIMIT {$offset}, {$rowCount}
       $queryString = " ( cc.sort_name LIKE '%$name%' OR cp.phone LIKE '%$name%' ) ";
     }
     else {
-      $cid = CRM_Utils_Request::retrieve('cid', 'CommaSeparatedIntegers', NULL, FALSE, NULL, 'GET');
+      $cid = CRM_Utils_Request::retrieveValue('cid', 'CommaSeparatedIntegers', NULL, FALSE, 'GET');
       if ($cid) {
         $queryString = " cc.id IN ( $cid )";
       }
@@ -480,9 +480,8 @@ LIMIT {$offset}, {$rowCount}
 
     if ($queryString) {
       $result = array();
-      $offset = CRM_Utils_Request::retrieve('offset', 'Integer', NULL, FALSE, 0, 'GET');
-      $rowCount = CRM_Utils_Request::retrieve('rowcount', 'Integer', NULL, FALSE, 20, 'GET');
-
+      $offset = (int) CRM_Utils_Request::retrieveValue('offset', 'Integer', 0, FALSE, 'GET');
+      $rowCount = (int) CRM_Utils_Request::retrieveValue('rowcount', 'Integer', 20, FALSE, 'GET');
       // add acl clause here
       list($aclFrom, $aclWhere) = CRM_Contact_BAO_Contact_Permission::cacheClause('cc');
       if ($aclWhere) {

```

Note that the (int) is required because of the way 0 is handled in retrieve function

For the 'name' retrieval it's more complex. Valid name variables include more than just alphanumeric. In my test there was a comma and a space (it works off sort_name) but also consider O'Connor. We probably need to filter things out ? Do we need a new filter

**Comment by seamuslee001** [comment]
@eileenmcnaughton possibly yes, i guess what i can do is for the moment send it back to string but yeh i think it would be good. Maybe we could allow for one space in the alphanumeric but hmmm its a difficult question

**Comment by eileenmcnaughton** [comment]
@seamuslee001 specifically stripping < & > seems good

File: CRM/Contact/Page/AJAX.php [hunk 1]
```diff
@@ -467,30 +467,21 @@ public static function getContactPhone() {
     $phoneTypes = CRM_Core_OptionGroup::values('phone_type', TRUE, FALSE, FALSE, NULL, 'name');
     $mobileType = CRM_Utils_Array::value('Mobile', $phoneTypes);
 
-    $name = CRM_Utils_Array::value('name', $_GET);
+    $name = CRM_Utils_Request::retrieveValue('name', 'String', NULL, FALSE, 'GET');
     if ($name) {
-      $name = CRM_Utils_Type::escape($name, 'String');
       $queryString = " ( cc.sort_name LIKE '%$name%' OR cp.phone LIKE '%$name%' ) ";
```
Comment (colemanw) [inline CRM/Contact/Page/AJAX.php]: I really don't think we should be concatenating string values from the url into a query. I think this line ought to be using `%1` `%2` etc placeholders and we should be feeding values into an array as the 2nd param to `executeQuery`.

**Comment by eileenmcnaughton** [comment]
@seamuslee001 I took a look at this - there are a few instance where you use retrieveValue instead of the existing CRM_Utils_Array::value + escape. For commaSeparatedInteger & Integer I think the validate action is solid enough not to also require escaping. 

However for the first one - ie. the one that is string  - the validate doesn't do much now we've backed off your original 'Alphanumeric' - making the change @colemanw suggested would work - although the way the hook has been jammed in that function it arguably could mess with it. Alternatively CRM_Core_DAO::composeQuery could be used to do the interpolation.

**Comment by eileenmcnaughton** [comment]
@seamuslee001 this needs an update by you per comments

**Comment by seamuslee001** [comment]
@eileenmcnaughton updated now

**Comment by eileenmcnaughton** [comment]
Thanks @seamuslee001 this is good to merge IMHO

**Comment by eileenmcnaughton** [comment]
@seamuslee001 oops I take it back - I'm getting a notice
screenshot 2018-10-25 17 06 19

**Comment by seamuslee001** [comment]
@eileenmcnaughton hopefully fixed that e-notice now

**Comment by eileenmcnaughton** [comment]
@seamuslee001 yep seems good

**Comment by seamuslee001** [comment]
Merging as per the tag

**Pull request merged**